### DISCUSSION TITLE
Custom Code to change certain lines of code for Kuka

### DISCUSSION THREAD

**Discussion by aimforkim** [visose/Robots#220] [Q&A] [answered]
Hello, 

I am currently using the robots plugin to write a toolpath (the Ptp and Lin movement) that will be pasted into workvisual.That would be branch {2} of the code output from the program.  The .dat file will be written separately inside workvisual (what would be {1} of the code output).

The current code output from the program need adjustments when I bring it to workvisual. I am manually changing them in workvisual, such as removing "$LOAD.CM = {X 7.48,Y 23.23,Z 132.2,A 0,B 0,C 0}" and muting "$VEL_AXIS[2] = 2.847". Is it possible to make these changes inside the program with the Custom Code component? Are there examples of using the custom code component (as I've been failing to understand what the input really requires)? Or would it make more sense that I create my own python script to parse the code and replace certain string? 

![image](https://user-images.githubusercontent.com/96722124/221157616-20ac3d7d-bb19-4b91-a446-3e7377016d41.png)

I hope this image clarifies my question. Thank you in advance!

**Comment by visose** [comment]
Those lines are there for a reason, it might be better to fix the root cause. The lines seem to be valid KRL syntax, but maybe Workvisual does not support all valid KRL. Maybe there's a better/alternative syntax to express that info that Workvisual understands. I'm happy to change the generated code to something that is compatible with Workvisual.

**Reply by aimforkim** [reply]
Thank you for the fast response. 
[https://github.com/visose/Robots/blob/1b14909eb19c3dfef930a1dbe13c4bac01232130/src/Robots/PostProcessors/KRLPostProcessor.cs](url) Yes, I believe this is what generates the program code. 

For our current workflow in workvisual, the Vel_Axis does not need to be defined [line 194-202]. There are times we set the tool, base, and velocity with BAS(). The $LOAD.CM shows up as an error, and I cannot quite identify where that is being written. These are the three things I can come up with that we commonly change.
![image](https://user-images.githubusercontent.com/96722124/221507198-46aa40b1-4575-41fd-9d17-be9cf4ce83b8.png)

> I'm happy to change the generated code to something that is compatible with Workvisual.

This would be great. However, the workflow is not fixed yet and there may be changes later in what will have to be written in Workvisual. Would it be possible to make a systemkuka_custom.cs that overrides from a custom post-processor (line 46), where we could make changes to the post-processor script according to how the workflow advances? I have not written a grasshopper component before, so I may be asking for a more complicated solution. If so, please let me know!

**Reply by visose** [reply]
The code of the post-processor is compiled into the plugin, so you cannot dynamically insert it like using a C# script in Grasshopper, if that's what you are thinking. To use your own custom post-processor, you would have to fork this repo, create a new one or modify the existing one, and compile a custom version of the plugin. 

I could allow others to insert their custom post processors in two ways:
- The easier way: I add some changes that include a post-processor like interface in the Robots library (there's a Nuget package). Someone else can create their own .NET assembly with a custom post processor that implements this interface. Then pass this custom postprocessor class into the CreateProgram component (Program class, in code). You could possibly use a c# scripting component for this, with an assembly reference to Robots.
- Lots more effort: Convert the postprocessors to use some templating language that gets processed at runtime, and can be modified by end users.
- Something in between those could be to use Python for the postprocessors.

I don't think these will happen anytime soon (unless this feature is sponsored). So your only option is to fork the repo and create your custom build of the Robots plugin.

**Reply by robertkovax** [reply]
Hi, I'd be also interested in customizing the code that is being generated. (I'd like to extend the capabilities of our Franka Emika robot). Was there any progress or experience on this issue since then? 

I did try to fork the repository, however I'm having problems with building and deploying the code. Is there a guide regarding the deployment workflow? Many thanks!

**Reply by visose** [reply]
Hi, I don't think there is going to be any changes on this for now. Regarding issues building the project feel free to open a new discussion.

**Comment by visose** [comment, marked as answer]
Closing this with the current answer: Custom Command can insert manufacturer-specific code at targets or declarations, and Custom Code can replace the whole generated program text, but they do not patch the compiled KUKA postprocessor itself. For systematic KRL output changes such as removing `$VEL_AXIS` or changing `$LOAD`, either post-process the generated text before feeding it to Custom Code, or create a custom/forked postprocessor.
